packages/libsql/src/lib/Category.ts

```diff
@@ -0,0 +1,417 @@
+    // in all other cases the snapshot is guaranteed to exist if there are any events in the stream
+    if (snapshot == null) return this.loadBatched(streamName, decode, fold, initial)
+    const [_, event] = snapshot
+    const decoded = decode(event)
+    // The snapshot type may have changed, in which case we should ignore it and load a fresh state
+    if (!decoded || !isOrigin(decoded)) return this.loadBatched(streamName, decode, fold, initial)
+    const version = event.index
+    const state = fold(initial, [decoded])
+    const span = trace.getActiveSpan()
+    span?.setAttributes({
+      [Tags.loaded_count]: 1,
+      [Tags.snapshot_version]: Number(version),
+    })
+    return [Token.create(version), state]
+  }
+
+  async reload<Event, State>(
+    streamName: Equinox.StreamName,
+    token: StreamToken,
+    decode: Decode<Event>,
+    fold: (state: State, events: Event[]) => State,
+    state: State,
+  ): Promise<[StreamToken, State]> {
+    let streamVersion = Token.version(token)
+    const startPos = streamVersion + 1n // Reading a stream uses {inclusive} positions
+    let batches = 0
+    let eventCount = 0
+    for await (const [version, events] of Read.loadForwardsFrom(
+      this.conn.read,
+      this.batchSize,
+      this.maxBatches,
+      streamName,
+      startPos,
+    )) {
+      state = fold(state, keepMap(events, decode))
+      streamVersion = streamVersion > version ? streamVersion : version
+      batches++
+      eventCount += events.length
+    }
+    trace.getActiveSpan()?.setAttributes({
+      [Tags.loaded_count]: eventCount,
+      [Tags.batches]: batches,
+    })
+    return [Token.create(streamVersion), state]
+  }
+
+  async syncRollingState(
+    category: string,
+    streamName: string,
+    token: StreamToken,
+    encodedEvent: IEventData<Format>,
+  ): Promise<GatewaySyncResult> {
+    const span = trace.getActiveSpan()
+    const etag = Token.snapshotEtag(token)
+    const index = Token.version(token)
+    const result = await this.conn.write.writeSnapshot(
+      category,
+      streamName,
+      encodedEvent,
+      index,
+      etag,
+    )
+    switch (result.type) {
+      case "ConflictUnknown":
+        span?.addEvent("Conflict")
+        return { type: "ConflictUnknown" }
+      case "Written": {
+        const token = Token.create(result.position, result.snapshot_etag)
+        return { type: "Written", token }
+      }
+    }
+  }
+
+  async sync(
+    category: string,
+    streamName: string,
+    token: StreamToken,
+    encodedEvents: IEventData<Format>[],
+    updateSnapshot?: (trx: Transaction) => Promise<void>,
+  ): Promise<GatewaySyncResult> {
+    const span = trace.getActiveSpan()
+    const streamVersion = Token.version(token)
+    const appendedTypes = new Set(encodedEvents.map((x) => x.type))
+    if (appendedTypes.size <= 10) {
+      span?.setAttribute(Tags.append_types, Array.from(appendedTypes))
+    }
+    const result = await this.conn.write.writeMessages(
```

Comment:
they are events man! (yes, I know why but still)

Reply:
They sure are, but I can also write commands into this table